cdap-api/src/main/java/io/cdap/cdap/api/app/ApplicationUpdateContext.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright © 2020 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.api.app;
+
+import com.google.gson.JsonSyntaxException;
+import io.cdap.cdap.api.Config;
+import io.cdap.cdap.api.artifact.ArtifactId;
+
+import java.lang.reflect.Type;
+import javax.annotation.Nullable;
+
+/**
+ * Context for updating Application configs. Provides helper methods for application to supporting operation on
+ * configs like upgrade.
+ */
+public interface ApplicationUpdateContext {
+
+  /**
+   * Get the old config as an object of the given type. The platform would perform the json deserialization based on
+   * the provided type. This is for the case where an application has the same/compatible/old config class. Application
+   * should decide on how they want to convert config from old to current type.
+   *
+   * @param configType type of the config platform should deserialize to.
+   * @return application config serialized to an object of given configType.
+   */
+  <C extends Config> C getConfig(Type configType) throws JsonSyntaxException;
+
+  /**
+   * Get the application configuration as json string.
+   */
+  String getConfigAsString();
+
+  /**
+   * Gets latest available artifact information for a given plugin type and name.
+   *
+   * @param pluginType the plugin type
+   * @param pluginName the plugin name
+   * @return artifact information for the plugin type and name.
+   *         returns null if no artifact for the plugin could be found.
+   */
+  @Nullable
+  ArtifactId getLatestPluginArtifact(String pluginType, String pluginName);
```

Comment:
Hum... I don't think it is always latest for the update purpose. 

Also, since the update is done through a similar cycle as deployment, meaning the `PluginConfigurer.usePlugin` will be called at some point during that process, in which it will determine the actual artifact to use. So what does the updateConfig need to do to make sure the `usePlugin` is selecting the right version during the actual deployment?

Reply:
Discussed offline. Added more filter parameters in getPluginArtifacts to let application handle how they want to consider plugins.